src/main/kotlin/io/span/libs/kms/KMSCSRSigner.kt

```diff
@@ -12,28 +13,70 @@ import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter
 import org.bouncycastle.pkcs.PKCS10CertificationRequest
 import org.slf4j.LoggerFactory
 import java.io.StringReader
+import java.math.BigInteger
+import java.security.Key
+import java.security.KeyFactory
+import java.util.Date
+import java.util.UUID
 import java.security.PublicKey
-import java.time.Duration
-import java.util.*
+import java.security.SecureRandom
+import java.security.spec.PKCS8EncodedKeySpec
 import javax.security.auth.x500.X500Principal
+import kotlin.math.log
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.days
 
 @Singleton
 class KMSBasedCSRSigner(
     private val caConfig: CertConfig,
     private val kmsClient: AWSKMSClient
 ) {
 
+    /**
+     * Given a key id of a private key present in KMS,
+     * generates a self-signed certificate for the given subject and validity.
+     * @param kmsKeyId - keyId of the private/public key pair present in KMS
+     * @param subjectCN - SubjectCN for the Root Certificate to be generated
+     * @param validity - the amount of time the certificate needs to be valid.
+     * @return DER encoded ASN.1 sequence representing the generated X509 Certificate
+     */
+    fun generateSelfSignedCert(kmsKeyId: String, subjectCN: String, validity: Duration): ByteArray {
+        val serialNumber = getRandomSerial()
+        val publicKey = getPublicKey(privateKeyId = kmsKeyId)
+        val now = Date()
+        val then = Date(now.time + validity.inWholeMilliseconds)
+        val subject = X500Principal("CN=$subjectCN")
```

Comment:
although the spec doesn't say you _have_ to have it you need an OID.1.3.6.1.4.1.37244.1.4=<some id> for the RCAC